testing/sd_bench/sd_bench_main.c

```diff
@@ -0,0 +1,476 @@
+}
+
+int read_test(int fd, sdb_config_t *cfg, uint8_t *block, int block_size)
+{
+  uint8_t *read_block;
+  uint64_t total_elapsed;
+  size_t total_blocks;
+  struct timespec start;
+  size_t num_blocks;
+  uint64_t max_read_time;
+  uint64_t read_time;
+  uint64_t elapsed;
+  struct timespec read_start;
+  size_t nread;
+
+  printf("\n");
+  printf("Testing Sequential Read Speed...\n");
+
+  read_block = (uint8_t *)malloc(block_size);
+
+  if (!read_block)
+    {
+      printf("Failed to allocate memory block\n");
+      return -1;
+    }
+
+  total_elapsed = 0.;
+  total_blocks = 0;
+  size_t *blocknumber = (unsigned int *)(void *) &read_block[0];
+
+  for (int run = 0; run < cfg->num_runs;  ++run)
+    {
+      start = get_abs_time();
+      num_blocks = 0;
+      max_read_time = 0;
+
+      while (get_elapsed_time_us(&start) < cfg->run_duration
+             && total_blocks + num_blocks < cfg->total_blocks_written)
+        {
+          read_start = get_abs_time();
+          nread = read(fd, read_block, block_size);
+          read_time = get_elapsed_time_us(&read_start);
+
+          if (read_time > max_read_time)
+            {
+              max_read_time = read_time;
+            }
+
+          if ((int)nread != block_size)
+            {
+              printf("Read error\n");
+              free(read_block);
+              return -1;
+            }
+
+          if (*blocknumber !=  total_blocks + num_blocks)
+            {
+              printf("Read data error at block: %d wrote:0x%04x read:0x%04x",
+                     (total_blocks + num_blocks),
+                     total_blocks + num_blocks, *blocknumber);
+            }
+
+          for (unsigned int i = sizeof(*blocknumber);
+              i < (block_size - sizeof(*blocknumber)); ++i)
+            {
+              if (block[i] != read_block[i])
+                {
+                  printf("Read data error at offset: %d wrote:0x%02x "
+                         "read:0x%02x", total_blocks + num_blocks + i,
+                         block[i], read_block[i]);
+                }
+            }
+
+          ++num_blocks;
+        }
+
+      elapsed = get_elapsed_time_us(&start);
+
+      if (num_blocks)
+        {
+          printf("  Run %2i: %8.1f KB/s, max read/verify time: %3.4f ms "
+                 "(%.1f KB/s)\n", run + 1,
+                 ts_to_kb(block_size * num_blocks, elapsed),
+                 max_read_time / 1e3,
+                 ts_to_kb(block_size, max_read_time));
+
+          total_elapsed += elapsed;
+          total_blocks += num_blocks;
+        }
+    }
+
+  printf("  Avg   : %8.1f KB/s, %3.3f MB and verified\n",
+         ts_to_kb(block_size * total_blocks, total_elapsed),
+         block_count_to_mb(total_blocks, block_size));
+
+  free(read_block);
+  return 0;
+}
+
+static void usage(void)
+{
+  printf("Test the speed of an SD card or mount point\n");
+  printf(CONFIG_TESTING_SD_BENCH_PROGNAME
+         ": [-b] [-r] [-d] [-k] [-s] [-u] [-v]\n");
```

Comment:
You removed unaligned. but -u is still here.

Personally I think unaligned is critical to the test. It demonstrates the performance hit with the FAT DMA allocator and the whole preflight fail - rerun cycle.

Reply:
That's a fair point and something I missed. I've added the option for switching between aligned and unaligned allocations.

In regards to the documentation, I'll do up a PR to fix up the error once this is approved / merged.